jira/resilientsession.py

```diff
@@ -1,119 +1,288 @@
+        resp (Optional[Response]): Response from Jira request
 
     Raises:
         JIRAError: If Response is None
         JIRAError: for unhandled 400 status codes.
-        JIRAError: for unhandled 200 status codes.
+
+    Returns:
+        TypeGuard[Response]: True if the passed in Response is all good.
     """
     request = kwargs.get("request", None)
-    # headers = kwargs.get('headers', None)
 
-    if r is None:
+    if resp is None:
         raise JIRAError(None, **kwargs)
 
-    if r.status_code >= 400:
-        error = ""
-        if r.status_code == 403 and "x-authentication-denied-reason" in r.headers:
-            error = r.headers["x-authentication-denied-reason"]
-        elif r.text:
-            try:
-                response = json.loads(r.text)
-                if "message" in response:
-                    # Jira 5.1 errors
-                    error = response["message"]
-                elif "errorMessages" in response and len(response["errorMessages"]) > 0:
-                    # Jira 5.0.x error messages sometimes come wrapped in this array
-                    # Sometimes this is present but empty
-                    errorMessages = response["errorMessages"]
-                    if isinstance(errorMessages, (list, tuple)):
-                        error = errorMessages[0]
-                    else:
-                        error = errorMessages
-                # Catching only 'errors' that are dict. See https://github.com/pycontribs/jira/issues/350
-                elif (
-                    "errors" in response
-                    and len(response["errors"]) > 0
-                    and isinstance(response["errors"], dict)
-                ):
-                    # Jira 6.x error messages are found in this array.
-                    error_list = response["errors"].values()
-                    error = ", ".join(error_list)
-                else:
-                    error = r.text
-            except ValueError:
-                error = r.text
+    if not resp.ok:
+        error = parse_error_msg(resp=resp)
+
         raise JIRAError(
             error,
-            status_code=r.status_code,
-            url=r.url,
+            status_code=resp.status_code,
+            url=resp.url,
             request=request,
-            response=r,
+            response=resp,
             **kwargs,
         )
-    # for debugging weird errors on CI
-    if r.status_code not in [200, 201, 202, 204]:
-        raise JIRAError(
-            status_code=r.status_code, request=request, response=r, **kwargs
-        )
-    # testing for the bug exposed on
-    # https://answers.atlassian.com/questions/11457054/answers/11975162
-    if (
-        r.status_code == 200
-        and len(r.content) == 0
-        and "X-Seraph-LoginReason" in r.headers
-        and "AUTHENTICATED_FAILED" in r.headers["X-Seraph-LoginReason"]
-    ):
-        pass
+
+    return True  # if no exception was raised, we have a valid Response
```

Comment:
I don't like returning True ....
if no exception => success

Reply:
this is required for the `TypeGuard` output type, which should only return True if it ensures that the input argument is the correct type (in this case a `Response`).

https://mypy.readthedocs.io/en/stable/type_narrowing.html#user-defined-type-guards